[quant][fx] Add pass in convert to fold quant-dequant sequence

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54860 [quant][fx] Add pass in convert to fold quant-dequant sequence**
* #54859 [quant][fx][pyper] Get first linear use of quantize_per_tensor for FQN

Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27390506](https://our.internmc.facebook.com/intern/diff/D27390506)

supriyar committed:
[quant][fx] Add pass in convert to fold quant-dequant sequence
Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cuspymd commented on torch/quantization/fx/quantize.py:
```diff
@@ -987,6 +987,34 @@ def load_arg(a):
+                quant_uses = list(node.users)
+                for user in quant_uses:
+                    if _is_dequant_op(user):
+                        quant_args = node.args
+                        float_tensor = quant_args[0]
```
How about moving out of the for loop of `quant_uses`?

supriyar committed:
Update on "[quant][fx] Add pass in convert to fold quant-dequant sequence"


Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27390506](https://our.internmc.facebook.com/intern/diff/D27390506)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -999,6 +999,34 @@ def load_arg(a):
+            In the case where the only consumer of quantize_per_tensor is a dequant op, we erase both
+            nodes from the graph, along with the qparams associated with quantize op.
+        """
+        def _is_dequant_op(node) -> bool:
+            if node.op == 'call_method' and node.target == "dequantize":
+                return True
+            else:
+                return False
```
would it be simple to inline it?

```
is_dequant = node.op == 'call_method' and node.target == "dequantize"
if is_dequant:
    ... do stuff ...
```

Asking because this file has a history of functions which are local to parent functions, and in those cases it's sometimes unclear which variables are passed in via args and which are from the parent function's scope.  If we need this to be a func, maybe we can move it to be a standalone function to make it easy to understand that all variables are passed in?

supriyar replied:
makes sense to inline it. I thought there would be more to this function when I first wrote it.

vkuzo reviewed (approved):
lgtm, perhaps would be good for @jerryzh168  to take a look as well?

supriyar committed:
Update on "[quant][fx] Add pass in convert to fold quant-dequant sequence"


Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27390506](https://our.internmc.facebook.com/intern/diff/D27390506)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Add pass in convert to fold quant-dequant sequence"


Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27390506](https://our.internmc.facebook.com/intern/diff/D27390506)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Add pass in convert to fold quant-dequant sequence"


Summary:
Currently we insert a quantize_per_tensor op when we encounter the quantizable input,
so if it has multiple uses and not all are quantizable then we need to add a dequantize op
before these ops.

In this pass - For a sequence of quantize_per_tensor - dequantize, we combine them
since it is a no-op.

Test Plan:
python test/test_quantization.py test_fold_quant_dequant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27390506](https://our.internmc.facebook.com/intern/diff/D27390506)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Accepting to unblock pyper, we can revert the change after a proper fix is landed

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -2058,6 +2058,45 @@ def forward(self, x):
+        m = prepare_fx(model, qconfig_dict)
+        m(torch.rand(5, 5))
+        m = convert_fx(m)
+        keys = m.state_dict().keys()
```
remove?

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -2058,6 +2058,45 @@ def forward(self, x):
+        m = convert_fx(m)
+        keys = m.state_dict().keys()
+        m(torch.randn(5, 5))
+        dequant = 0
+        quant = 0
+        for n in m.graph.nodes:
+            if n.op == "call_method" and n.target == "dequantize":
+                dequant = dequant + 1
+            if n.op == "call_function" and n.target == torch.quantize_per_tensor:
+                quant = quant + 1
+        self.assertEqual(dequant, 1)
+        self.assertEqual(quant, 1)
```
we can also use self.checkGraphModuleNodes here I think, please take a look at other tests to see how to use it
